packages/eslint-plugin/src/rules/no-array-delete.ts

```diff
@@ -0,0 +1,90 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import type * as ts from 'typescript';
+
+import {
+  createRule,
+  getConstrainedTypeAtLocation,
+  getParserServices,
+} from '../util';
+
+type MessageId = 'noArrayDelete';
+
+export default createRule<[], MessageId>({
+  name: 'no-array-delete',
+  meta: {
+    fixable: 'code',
+    type: 'problem',
+    docs: {
+      description: 'Disallow using the `delete` operator on array values',
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      noArrayDelete:
+        'Using the `delete` operator with an array expression is unsafe. Use `array.splice()` instead.',
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    function isUnderlyingTypeArray(type: ts.Type): boolean {
+      const predicate = (t: ts.Type): boolean =>
+        checker.isArrayType(t) || checker.isTupleType(t);
+
+      if (type.isUnion()) {
+        return type.types.every(predicate);
+      }
+
+      if (type.isIntersection()) {
+        return type.types.some(predicate);
+      }
+
+      return predicate(type);
+    }
+
+    return {
+      'UnaryExpression[operator="delete"]'(
+        node: TSESTree.UnaryExpression,
+      ): void {
+        if (node.argument.type !== AST_NODE_TYPES.MemberExpression) {
+          return;
+        }
+
+        const type = getConstrainedTypeAtLocation(
+          services,
+          node.argument.object,
+        );
+
+        if (!isUnderlyingTypeArray(type)) {
+          return;
+        }
+
+        context.report({
+          node,
+          messageId: 'noArrayDelete',
+          fix(fixer): TSESLint.RuleFix | null {
```

Comment:
[Bug] Err, why was this switched from `suggestion` to `fix`? As undesirable as the reported behavior is, switching from `delete` to `.splice()` changes code behavior and so can't be a `fix`.

Reply:
hmm
actually, I have no idea why I did this haha
for some reason, I remember us talking about changing it to a fix, but I'm probably mixing it with another PR 😓 

fixed :)